### ISSUE TITLE
[nio] think about API too offload protbuf decoding onto a separate thread

### ISSUE THREAD

**Issue by weissi** [grpc/grpc-swift#626]
In the `nio` version, protobuf decoding is currently always inline on the NIO EventLoop threads. This might become a problem in the future and the only mitigation might be to offload that to background threads.

We should investigate if it is possible to get this in before the `1.0.0` release.

CC @glbrntt

**Comment by MrMage** [comment]
I guess this would apply to encoding response payloads as well. In my experience, protobuf decoding and encoding takes very little time (<1ms) for reasonably-sized payloads, not making a practical difference with regards to latency. Let me know if you are seeing different results for your workloads.

Also, we plan to expose only very high-level APIs for the 1.0 release. So even if this feature wouldn't make it into 1.0 (we are trying to keep scope small for a 1.0 release before EOY), we should be able to add this functionality in a (purely additive) minor release without any API-breaking changes.

**Comment by weissi** [comment]
@MrMage I agree with all of that. If we believe we can add that without breaking API, then I think we can close this ticket. We don't need it right this second but I wanted to be sure we wouldn't need to break API to add this (if required).

**Comment by MrMage** [comment]
👍 

Let's keep this open as a feature request for post-1.0. For 1.0 a goal is specifically to have such a small API surface to allow additions like this without breaking things.
